app/domain/authentication/resource_restrictions/extract_resource_restrictions.rb

```diff
@@ -0,0 +1,120 @@
+        create_resource_restrictions_object
+
+        @logger.debug(LogMessages::Authentication::ExtractedResourceRestrictions.new(resource_restrictions.names))
+
+        resource_restrictions
+      end
+
+      private
+
+      def fetch_resource_annotations
+        resource_annotations
+      end
+
+      def resource_annotations
+        @resource_annotations ||=
+          host_resource.annotations.each_with_object({}) do |annotation, result|
+            annotation_values = annotation.values
+            value = annotation_values[:value]
+            next if value.blank?
+            result[annotation_values[:name]] = value
+          end
+      end
+
+      def host_resource
+        # Validate role exists, otherwise getting role annotations return empty hash.
+        role_id = @role_class.roleid_from_username(@account, @host_name)
+        host_resource = @resource_class[role_id]
+
+        raise Errors::Authentication::Security::RoleNotFound, role_id unless host_resource
+
+        host_resource
+      end
+
+      def extract_resource_restrictions_from_annotations
+        resource_restrictions_hash
+      end
+
+      def resource_restrictions_hash
+        @resource_restrictions_hash ||=
+          resource_annotations.each_with_object({}) do |(annotation_name, annotation_value), resource_restrictions_hash|
+            add_restriction_to_hash(annotation_name, annotation_value, resource_restrictions_hash)
+          end
+      end
+
+      def add_restriction_to_hash(annotation_name, annotation_value, resource_restrictions_hash)
+        restriction_name, is_general_restriction = get_restriction_from_annotation(annotation_name)
+
+        return unless restriction_name
+
+        # General restriction should not override existing restriction
+        return if is_general_restriction && resource_restrictions_hash.include?(restriction_name)
+
+        @logger.debug(LogMessages::Authentication::RetrievedAnnotationValue.new(annotation_name))
+
+        resource_restrictions_hash[restriction_name] = annotation_value
+      end
+
+      # Parses the given annotation name and tries to extract a restriction from it.
+      # A restriction can have 2 formats:
+      # 1. "<Authenticator name>/<Restriction name>"
+      # 2. "<Authenticator name>/<Service ID>/<Restriction name>"
+      #
+      # The first format is a general restriction.
+      # This means it is relevant to all services of this authenticator.
+      #
+      # The second format is specific to the specified service ID.
+      # This means all other services will ignore it.
+      # Moreover, if both formats are found for the same restriction, the specific one will be used.
+      #
+      # This function returns the restriction name, if found (nil otherwise) and a
+      # boolean indicating if it is a general restriction or not.
+      def get_restriction_from_annotation(annotation_name)
+        annotation_name.match(authenticator_prefix_regex) do |prefix_match|
```

Comment:
what if there was something like `authn-gcp-sigal` would this still be valid?

Reply:
This class does not validate anything (as its name imply).
So any restriction is valid.
It might not be found, because part of this logic is to extract only the relevant resource restrictions.
So for `authn-gcp`, if you define an annotation `authn-gcp-sigal/project-id`, it won't be found.
But if you define `authn-gcp/sigal-project-id`, it will find the `sigal-project-id` resource restriction (whatever that means... 😉 )

Reply:
isn't it the project id param for a Sigal Google Cloud Platform machine?